### PR TITLE
Add the ability to only compile and install dependencies, but not build the binaries

### PR DESCRIPTION
**Pull request by madhusudancs**
Linking is terribly slow. Sometimes I don't actually care about the binary, but I only want to make sure that the code compiles. This is especially true while refactoring. These changes enable just building the code without actually linking the final binaries.

cc @mikedanese @wojtek-t

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33731)

### CODE REVIEW

File: hack/lib/golang.sh [hunk 1]
```diff
@@ -536,25 +536,42 @@ kube::golang::build_binaries_for_platform() {
...
 
-    # `go test -c` below directly builds the binary. It builds the packages,
-    # but it never installs them. `go test -i` only installs the dependencies
-    # of the test, but not the test package itself. So neither `go test -c`
-    # nor `go test -i` installs, for example, test/e2e.a. And without that,
-    # doing a staleness check on k8s.io/kubernetes/test/e2e package always
-    # returns true (always stale). And that's why we need to install the
-    # test package.
-    go install "${goflags[@]:+${goflags[@]}}" \
+    if [[ ! -n ${use_go_build:-} ]]; then
```
Comment (thockin) [inline hack/lib/golang.sh]: use `[[ -z` rather than `! -n`, and wrap `${use_go_build:-}` in quotes.

File: hack/lib/golang.sh [hunk 2]
```diff
@@ -536,25 +536,42 @@ kube::golang::build_binaries_for_platform() {
...
+      # doing a staleness check on k8s.io/kubernetes/test/e2e package always
+      # returns true (always stale). And that's why we need to install the
+      # test package.
+      go install "${goflags[@]:+${goflags[@]}}" \
+          -gcflags "${gogcflags}" \
+          -ldflags "${goldflags}" \
+          "${testpkg}"
+    fi
+
+    test_install=false
```
Comment (thockin) [inline hack/lib/golang.sh]: This deserves at least as much comment as above.  I'm having a hard time figuring it out.

File: hack/lib/golang.sh [hunk 3]
```diff
@@ -536,25 +536,42 @@ kube::golang::build_binaries_for_platform() {
...
+    for goflag in "${goflags[@]:+${goflags[@]}}"; do
+      if [[ "${goflag}" == "-i" ]]; then
```
Comment (thockin) [inline hack/lib/golang.sh]: Why would goflags include -i ?  Who puts that there?

File: hack/lib/golang.sh [hunk 4]
```diff
@@ -632,14 +650,18 @@ kube::golang::build_binaries() {
...
 
-    if [[ ${#targets[@]} -eq 0 ]]; then
-      targets=("${KUBE_ALL_TARGETS[@]}")
+    if [[ ${#all_targets[@]} -eq 0 ]]; then
+      all_targets=("${KUBE_ALL_TARGETS[@]}")
     fi
 
+    # Dedup the `all_targets` list.
+    # Credit: http://stackoverflow.com/a/13648438
+    targets=($(echo "${all_targets[@]}" | tr ' ' '\n' | sort -u | tr '\n' ' '))
```
Comment (thockin) [inline hack/lib/golang.sh]: under what conditions would we have dups?

File: hack/lib/golang.sh [hunk 5]
```diff
@@ -662,8 +684,12 @@ kube::golang::build_binaries() {
       fi
     fi
 
-    # First build the toolchain before building any other targets
-    kube::golang::build_kube_toolchain
+    # First install the toolchain before building any other targets.
+    # We are only using this toolchain when installing packages, so
+    # it is not necessary to install them when we are just building
+    if [[ ! -n ${use_go_build:-} ]]; then
```
Comment (thockin) [inline hack/lib/golang.sh]: -z and quotes

**Comment by dims** [comment]
@madhusudancs Do we need this for 1.5? (I see a 1.5 milestone)

**Comment by madhusudancs** [comment]
@dims no. My plan was to address the comments after the code freeze anyway.

**Pull request closed without merging**